File: _may_messenger_mobile_app/lib/data/services/search_service.dart
import 'package:dio/dio.dart';
import '../models/user_model.dart';
import '../models/search_result_model.dart';
import '../models/chat_model.dart';

class SearchService {
  final Dio _dio;

  SearchService(this._dio);

  Future<List<User>> searchUsers(String query) async {
    try {
      final response = await _dio.get(
        '/api/users/search',
        queryParameters: {'query': query},
      );
      
      if (response.data is List) {
        return (response.data as List)
            .map((json) => User.fromJson(json))
            .toList();
      }
      return [];
    } catch (e) {
      print('[SearchService] Error searching users: $e');
      rethrow;
    }
  }

  Future<List<Chat>> searchChats(String query) async {
    try {
      final response = await _dio.get(
        '/api/chats/search',
        queryParameters: {'query': query},
      );
      
      if (response.data is List) {
        return (response.data as List)
            .map((json) => Chat.fromJson(json))
            .toList();
      }
      return [];
    } catch (e) {
      print('[SearchService] Error searching chats: $e');
      rethrow;
    }
  }

  Future<List<MessageSearchResult>> searchMessages(String query) async {
    try {
      final response = await _dio.get(
        '/api/messages/search',
        queryParameters: {'query': query},
      );
      
      if (response.data is List) {
        return (response.data as List)
            .map((json) => MessageSearchResult.fromJson(json))
            .toList();
      }
      return [];
    } catch (e) {
      print('[SearchService] Error searching messages: $e');
      rethrow;
    }
  }
}

File: _may_messenger_web_client/src/services/searchService.ts
/**
 * Search Service - поиск пользователей и сообщений
 */

import { apiClient } from '../api/apiClient';

export interface User {
  id: string;
  displayName: string;
  phoneNumber: string;
  isOnline?: boolean;
  lastSeenAt?: string;
}

export interface MessageSearchResult {
  messageId: string;
  chatId: string;
  chatTitle: string;
  messageContent: string;
  senderName: string;
  createdAt: string;
}

class SearchService {
  /**
   * Поиск пользователей по запросу
   * @param query - поисковый запрос
   * @param contactsOnly - искать только среди контактов
   */
  async searchUsers(query: string, contactsOnly: boolean = false): Promise<User[]> {
    try {
      const response = await apiClient.get('/api/users/search', {
        params: {
          query,
          contactsOnly,
        },
      });

      if (Array.isArray(response.data)) {
        return response.data as User[];
      }

      return [];
    } catch (error) {
      console.error('[SearchService] Error searching users:', error);
      throw error;
    }
  }

  /**
   * Поиск сообщений по запросу
   * @param query - поисковый запрос
   */
  async searchMessages(query: string): Promise<MessageSearchResult[]> {
    try {
      const response = await apiClient.get('/api/messages/search', {
        params: { query },
      });

      if (Array.isArray(response.data)) {
        return response.data as MessageSearchResult[];
      }

      return [];
    } catch (error) {
      console.error('[SearchService] Error searching messages:', error);
      throw error;
    }
  }

  /**
   * Комбинированный поиск (пользователи + сообщения)
   */
  async searchAll(query: string): Promise<{
    users: User[];
    messages: MessageSearchResult[];
  }> {
    try {
      const [users, messages] = await Promise.all([
        this.searchUsers(query),
        this.searchMessages(query),
      ]);

      return { users, messages };
    } catch (error) {
      console.error('[SearchService] Error in combined search:', error);
      throw error;
    }
  }
}

// Экспортируем синглтон
export const searchService = new SearchService();

